src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -255,8 +256,57 @@ private void RecordDependency(CallGraphNode callerKey, CallGraphNode calledKey,
         /// </summary>
         public List<ImmutableArray<CallGraphNode>> GetCallCycles()
         {
-            // ToDo: need to implement a robust algorithm for finding cycles
-            throw new NotImplementedException();
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
```

Comment:
If the node is not in the graph, does this indicate an error, or is it that library functions aren't tracked in the graph, so it's valid for nodes to call other nodes that don't exist?

Reply:
All the nodes have to be in the graph, or this will error. I added logic to make sure that all the nodes are in the graph.

Reply:
Sorry, I don't see the added logic?

Reply:
Sorry, looks like that didn't come into play until later: https://github.com/microsoft/qsharp-compiler/blob/sccarda/JohnsonsAlgorithm/src/QsCompiler/Transformations/CallGraphWalker.cs#L405